File: 11-graph/00200-Number_Of_Islands/00200-number_of_islands.cpp
// 200: Number Of Islands
// https://leetcode.com/problems/number-of-islands/

#include <vector>
#include <iostream>

using namespace std;


class Solution {
public:
    // SOLUTION
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        int result = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j, m, n);
                    result++;
                }
            }
        }

        return result;
    }
private:
    void dfs(vector<vector<char>>& grid, int i, int j, int m, int n) {
        if (
            i < 0 ||
            i >= m ||
            j < 0 ||
            j >= n ||
            grid[i][j] == '0'
        ) {
            return;
        }

        grid[i][j] = '0';

        dfs(grid, i - 1, j, m, n);
        dfs(grid, i + 1, j, m, n);
        dfs(grid, i, j - 1, m, n);
        dfs(grid, i, j + 1, m, n);
    }
};


int main() {
    Solution o;

    // INPUT
    vector<vector<char>> grid = {
        {'1','1','1','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','0'},
        {'0','0','0','0','0'}
    };

    // OUTPUT
    auto result = o.numIslands(grid);
    cout << result << endl;

    return 0;
}

File: 11-graph/00200-Number_Of_Islands/00200-number_of_islands.go
// 200: Number Of Islands
// https://leetcode.com/problems/number-of-islands/

package main

import "fmt"

func dfs(grid [][]byte, i, j int) {
	if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[i]) || grid[i][j] != '1' {
		return
	}

	grid[i][j] = '2'
	dfs(grid, i+1, j)
	dfs(grid, i-1, j)
	dfs(grid, i, j+1)
	dfs(grid, i, j-1)
}

// SOLUTION
func numIslands(grid [][]byte) int {
	result := 0

	for i, row := range grid {
		for j, point := range row {
			if point == '1' {
				result++
				dfs(grid, i, j)
			}
		}
	}

	return result
}

func main() {
	// INPUT
	grid := [][]byte{
		{'1', '1', '1', '1', '0'},
		{'1', '1', '0', '1', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '0', '0', '0'},
	}

	// OUTPUT
	result := numIslands(grid)
	fmt.Println(result)
}
